packages/formatter-markdown/src/formatter.ts

```diff
@@ -0,0 +1,224 @@
+
+    private language: string = '';
+
+    private getMessage(key: MessageName, substitutions?: string | string[]) {
+        return getMessageFormatter(key, this.language, substitutions);
+    }
+
+
+    public async format(problems: Problem[], options: FormatterOptions = {}) {
+
+        this.language = options.language!;
+        const target = options.target || '';
+        const result = new AnalysisResult(target, options);
+        const categoryList: string[] = getCategoryList(options.resources);
+
+        categoryList.forEach((category) => {
+            result.addCategory(category, this.language);
+        });
+
+        problems.forEach((message) => {
+            result.addProblem(message, this.language);
+        });
+
+        /* istanbul ignore if */
+        if (options.resources) {
+            options.resources.hints.forEach((hintConstructor) => {
+                const categoryName: string = hintConstructor.meta.docs!.category!;
+                const hintId: string = hintConstructor.meta.id;
+
+                const category: CategoryResult = result.getCategoryByName(categoryName)!;
+                const hint: HintResult | undefined = category.getHintByName(hintId);
+
+                if (!hint) {
+                    category.addHint(hintId, 'pass');
+                }
+            });
+        }
+
+        try {
+            if (!options.noGenerateFiles) {
+                result.percentage = 100;
+                result.id = Date.now().toString();
+
+                await createLanguageFile(this.language);
+
+                const markdown = this.createMarkdown(result);
+
+                await removeLanguageFile();
+
+                // We save the result with the friendly target name
+                const name = target.replace(/:\/\//g, '-')
+                    .replace(/:/g, '-')
+                    .replace(/\./g, '-')
+                    .replace(/\//g, '-')
+                    .replace(/[?=]/g, '-query-')
+                    .replace(/-$/, '');
+
+                const destDir = options.output || path.join(cwd(), 'hint-report');
+
+                const destination = path.join(destDir, `${name}.md`);
+
+                await fs.outputFile(destination, markdown);
+            }
+
+            return result;
+        } catch (err) {
+            logger.error(err);
+
+            throw err;
+        }
+    }
+
+    /**
+     * Creates the markdown report of the webhint scan.
+     * @param result The webhint scan result.
+     */
+    /* istanbul ignore next [too hard to test] */
+    private createMarkdown(result: AnalysisResult) {
+        let markdown = '';
+
+        markdown += MarkdownHelpers.createHeader(`Webhint Report - ${result.date}`, HeaderCount.Title);
+        markdown += MarkdownHelpers.newLine;
+        markdown += `${this.getMessage('hints')}: ${result.hintsCount}`;
+        markdown += MarkdownHelpers.newLine;
+
+        result.categories.forEach((category) => {
```

Comment:
What do you think about refactoring this a bit to make it easier to read?
Maybe replace each `forEach` with a `map` so it can return the content for each one and then you can add them to `markdown` after a `[].join(MardownHelpers.newlLine)` or something similar?

Reply:
@molant I've changed this to use map but I'm not sure if it is right or not.